Feature-002(Crear la entidad User y la enumeración role)

Christian-Cachero opened the pull request:
[Adicionalmente se agregaron los correspondientes mappers.](https://trello.com/c/gRlPIoU6)

Por favor revisar y darme su opinión.

Christian-Cachero committed:
Se completó las tareas de la feature 002, adicionalmente se agregó los mappers correspondientes.

Christian-Cachero committed:
Se completó las tareas de la feature 002, adicionalmente se agregó los mappers correspondientes.

 Se corrigió el error señalado.
 Y unos pequeños detalles

SergioZF09 reviewed (changes requested):
Ahorita revisé los cambios que hiciste Christian, y te comento una observación en UserRequestDTO y UserResponseDTO tiene los atributos description y price, y en UserEntity no tiene esos atributos, y no sé si pueda o no dar incovenientes por ejemplo para crear el usuario.

Christian-Cachero commented:
> Ahorita revisé los cambios que hiciste Christian, y te comento una observación en UserRequestDTO y UserResponseDTO tiene los atributos description y price, y en UserEntity no tiene esos atributos, y no sé si pueda o no dar incovenientes por ejemplo para crear el usuario.

Hola, tenes razón no me abre dado cuenta ahora lo corrijo.

Gracias!

Christian-Cachero committed:
Merge remote-tracking branch 'origin/feature-002-entidades-y-dtos' into feature-002-entidades-y-dtos

# Conflicts:
#	src/main/java/com/equipo5/backend/model/dtos/request/UserRequestDTO.java
#	src/main/java/com/equipo5/backend/model/dtos/response/BookingResponseDTO.java
#	src/main/java/com/equipo5/backend/model/dtos/response/ServiceEntityResponseDTO.java
#	src/main/java/com/equipo5/backend/model/dtos/response/UserResponseDTO.java

Christian-Cachero committed:
Merge remote-tracking branch 'origin/feature-002-entidades-y-dtos' into feature-002-entidades-y-dtos

# Conflicts:
#	src/main/java/com/equipo5/backend/model/dtos/request/UserRequestDTO.java
#	src/main/java/com/equipo5/backend/model/dtos/response/BookingResponseDTO.java
#	src/main/java/com/equipo5/backend/model/dtos/response/ServiceEntityResponseDTO.java
#	src/main/java/com/equipo5/backend/model/dtos/response/UserResponseDTO.java

Christian-Cachero committed:
Merge remote-tracking branch 'origin/feature-002-entidades-y-dtos' into feature-002-entidades-y-dtos

Christian-Cachero commented:
Ya corregí mis errores y los mergee a mi propia rama para que hicieren efecto directamente en este PR por favor revisarlos y comentame.

SergioZF09 reviewed (commented):
Sí los corregiste muy bien. Y tengo otra observación, en BookingRequestDTO no incluiste las relaciones con pets, services y owners, si no es necesario entonces está bien como está actualmente.

Christian-Cachero commented:
> Sí los corregiste muy bien. Y tengo otra observación, en BookingRequestDTO no incluiste las relaciones con pets, services y owners, si no es necesario entonces está bien como está actualmente.

Hmmm... tiene sentido, mañana me fijo, voy a hacer unos cambios para que se generen las fechas solas creo y también veré eso otras variables.

matiasnm commented:
**Recomiendo:**

1. Agregar validaciones a todos los campos de los DTO. ej:

> import jakarta.validation.constraints.NotBlank ;
> import jakarta.validation.constraints.NotNull;
> 
> public record ServiceEntityRequestDTO(
> NotBlank (message = "Incomplete attribute: 'name'.") String name,
> NotBlank (message = "Incomplete attribute: 'description'.") String description,
> NotNull (message = "Incomplete attribute: 'price'.") Double price) {
> }

2. Problemas entre Tablas y Nombre de Entidades. Ejemplo: en el esquema ER tenemos _table users_, pero en Spring la entidad es _UserEntity_

Las tablas suelen nombrarse en minusculas y las etidades comienzan en mayusculas, Spring automaticamente las convierte a minusculas y busca la tabla adecuada, pero deben tener el mismo nombre! Solucion, forzar todos los nombres de tabla para mas control agregando a cada entidad, por ejemplo:

> @table(name = "users")

3. Los campos de las entidades deben conicidir con los campos en el esquema ER, ejemplo, _long id_ nunca va a funcionar si la tabla usa _id_user_. La soucion es usar en cada campo @column(name = "id_user")

Ejemplo para name:

> @column(name = "name_user", nullable = false, length = 80)
> private String name;

4. Agregar import jakarta.validation.constraints.*; y usar en todos los campos de las entidades con @NotNull o @NotEmpty cuando sea necesario.

@NotNull → cuando sólo importa que el valor exista (ej: id obligatorio, booleano, fecha).
@NotEmpty → cuando además de existir, no puede estar vacío (ej: lista de items en una orden).
@NotBlank → cuando es String y además no puede ser solo espacios en blanco (ej: nombre de usuario, email).

5. Campo _created_at_user_ y _updated_at_user_

Usa este codigo para llenar ambos campos automaticamente, agrego los campos y nombres correctos de la tabla en el ejemplo:

> @column(name = "created_at_user", updatable = false)
> @NotNull
> private Instant createdAt;
> 
> @column(name = "updated_at_user")
> @NotNull
> private Instant updatedAt;
> 
> @PreUpdate
> @PrePersist
> public void updateTimeStamps() {
> updatedAt = Instant.now();
> if (createdAt == null) {
> createdAt = Instant.now();
> }
> }

Christian-Cachero commented:
> **Recomiendo:**
> 
>     1. Agregar validaciones a todos los campos de los DTO. ej:
> 
> 
> > import jakarta.validation.constraints.NotBlank ;
> > import jakarta.validation.constraints.NotNull;
> > public record ServiceEntityRequestDTO(
> > NotBlank (message = "Incomplete attribute: 'name'.") String name,
> > NotBlank (message = "Incomplete attribute: 'description'.") String description,
> > NotNull (message = "Incomplete attribute: 'price'.") Double price) {
> > }
> 
>     2. Problemas entre Tablas y Nombre de Entidades. Ejemplo: en el esquema ER tenemos _table users_, pero en Spring la entidad es _UserEntity_
> 
> 
> Las tablas suelen nombrarse en minusculas y las etidades comienzan en mayusculas, Spring automaticamente las convierte a minusculas y busca la tabla adecuada, pero deben tener el mismo nombre! Solucion, forzar todos los nombres de tabla para mas control agregando a cada entidad, por ejemplo:
> 
> > @table(name = "users")
> 
>     3. Los campos de las entidades deben conicidir con los campos en el esquema ER, ejemplo, _long id_ nunca va a funcionar si la tabla usa _id_user_. La soucion es usar en cada campo @column(name = "id_user")
> 
> 
> Ejemplo para name:
> 
> > @column(name = "name_user", nullable = false, length = 80)
> > private String name;
> 
>     4. Agregar import jakarta.validation.constraints.*; y usar en todos los campos de las entidades con @NotNull o @notempty cuando sea necesario.
> 
> 
> @NotNull → cuando sólo importa que el valor exista (ej: id obligatorio, booleano, fecha). @notempty → cuando además de existir, no puede estar vacío (ej: lista de items en una orden). @notblank → cuando es String y además no puede ser solo espacios en blanco (ej: nombre de usuario, email).
> 
>     5. Campo _created_at_user_ y _updated_at_user_
> 
> 
> Usa este codigo para llenar ambos campos automaticamente, agrego los campos y nombres correctos de la tabla en el ejemplo:
> 
> > @column(name = "created_at_user", updatable = false)
> > @NotNull
> > private Instant createdAt;
> > @column(name = "updated_at_user")
> > @NotNull
> > private Instant updatedAt;
> > @PreUpdate
> > @PrePersist
> > public void updateTimeStamps() {
> > updatedAt = Instant.now();
> > if (createdAt == null) {
> > createdAt = Instant.now();
> > }
> > }

lo de las validaciones es para mas tarde cuando la aplicación tenga todo lo necesario para funcionar y probarla.

en cuanto a los nombres de las tablas concuerdo para algunas entidades.

tambien lo de @PreUpdate @PrePersist estaba pensado agregarlo hoy por que es algo 100% necesario en esta etapa.

Christian-Cachero commented:
> **Recomiendo:**
> 
>     1. Agregar validaciones a todos los campos de los DTO. ej:
> 
> 
> > import jakarta.validation.constraints.NotBlank ;
> > import jakarta.validation.constraints.NotNull;
> > public record ServiceEntityRequestDTO(
> > NotBlank (message = "Incomplete attribute: 'name'.") String name,
> > NotBlank (message = "Incomplete attribute: 'description'.") String description,
> > NotNull (message = "Incomplete attribute: 'price'.") Double price) {
> > }
> 
>     2. Problemas entre Tablas y Nombre de Entidades. Ejemplo: en el esquema ER tenemos _table users_, pero en Spring la entidad es _UserEntity_
> 
> 
> Las tablas suelen nombrarse en minusculas y las etidades comienzan en mayusculas, Spring automaticamente las convierte a minusculas y busca la tabla adecuada, pero deben tener el mismo nombre! Solucion, forzar todos los nombres de tabla para mas control agregando a cada entidad, por ejemplo:
> 
> > @table(name = "users")
> 
>     3. Los campos de las entidades deben conicidir con los campos en el esquema ER, ejemplo, _long id_ nunca va a funcionar si la tabla usa _id_user_. La soucion es usar en cada campo @column(name = "id_user")
> 
> 
> Ejemplo para name:
> 
> > @column(name = "name_user", nullable = false, length = 80)
> > private String name;
> 
>     4. Agregar import jakarta.validation.constraints.*; y usar en todos los campos de las entidades con @NotNull o @notempty cuando sea necesario.
> 
> 
> @NotNull → cuando sólo importa que el valor exista (ej: id obligatorio, booleano, fecha). @notempty → cuando además de existir, no puede estar vacío (ej: lista de items en una orden). @notblank → cuando es String y además no puede ser solo espacios en blanco (ej: nombre de usuario, email).
> 
>     5. Campo _created_at_user_ y _updated_at_user_
> 
> 
> Usa este codigo para llenar ambos campos automaticamente, agrego los campos y nombres correctos de la tabla en el ejemplo:
> 
> > @column(name = "created_at_user", updatable = false)
> > @NotNull
> > private Instant createdAt;
> > @column(name = "updated_at_user")
> > @NotNull
> > private Instant updatedAt;
> > @PreUpdate
> > @PrePersist
> > public void updateTimeStamps() {
> > updatedAt = Instant.now();
> > if (createdAt == null) {
> > createdAt = Instant.now();
> > }
> > }

lo de las validaciones es para mas tarde cuando la aplicación tenga todo lo necesario para funcionar y probarla.

en cuanto a los nombres de las tablas concuerdo para algunas entidades.

también lo de @PreUpdate @PrePersist estaba pensado agregarlo hoy por que es algo 100% necesario en esta etapa.

Christian-Cachero committed:
hice las modificaciones pertinentes.

 se aplicaron validaciones a todos los RequestDTOs, ya que son los que reciben los datos desde el cliente.

 No creo que será necesario implementar la sugerencia 3 porque estamos usando un ORM que es el que se encarga de hacer las tablas en la BD y por ende no va a haber un conflicto.

SergioZF09 reviewed (approved):
Por mi parte está todo correcto.

matiasnm commented:
Barbaro, me bajo la rama y arranco con los test!
